neonvm: Add support for virtio-mem

sharnoff opened the pull request:
Adapted from #705.

This implementation is focused on making virtio-mem a drop-in replacement for our existing DIMM hotplug -based memory scaling.

As such, the externally visible differences are:

1. VirtualMachines have a new optional field: `.spec.guest.memoryProvider`
2. VirtualMachines have a new optional field: `.status.memoryProvider`
3. VirtualMachine `.status.memorySize` changes in smaller increments if using virtio-mem.

This change introduces the concept of a "memory provider", which is either "DIMMSlots" or "VirtioMem".

If a VM is created without setting the memory provider in the spec, the controller will use the default one, which is specified by its `--default-memory-provider` flag.
This will only persist for the lifetime of the runner pod. If the VM is restarted, then it may change to a new default.

Additionally, because our usage of virtio-mem has 8MiB block sizes, we actually *don't* default to virtio-mem if the VM was created with some odd memory slot size that isn't divisible by 8MiB.

The memory provider of the current runner pod is stored in the new status field. For old VMs created before this change, the status field is set to DIMMSlots iff the pod name is set without the status field.

Also, we allow mutating `.spec.guest.memoryProvider` IFF it is being set equal to the current `.status.memoryProvider`, in case we we want old VMs to keep their dimm slots after switching to virtio-mem by default.

---

Changes to split out from this PR:

1. #971
2. #973
3. #974

On top of those, there's a couple places where code was indented that should probably be split into its own commit so it can be properly added to .git-blame-ignore-revs :)

---

Remaining testing for this PR before merging:

- [x] Test that backwards compatibility works as expected
   - [x] Old VMs are updated while running as described above
   - [x] Old VMs switch to the new default on restart
- [x] Test rollback safety: that it's ok to go between the last release and this PR (with `--default-memory-provider=DIMMSlots`).
   - Note: If `--default-memory-provider=VirtioMem` when this is released, we cannot make it rollback-safe.
- [ ] ~~Test that updating `.spec.guest.memoryProvider` works as expected~~

---

Future/follow-up work:

1. #981
2. Switch to virtio-mem by default
3. Remove support for DIMM hotplug
4. Move from representing memory as "memory slots" towards the implementation
5. Fix the raciness issue with migration target pod creation.

Omrigan reviewed (commented):
Just a couple thoughts

sharnoff commented:
Cleaned up the commits in this PR -- added single commits for #973 and #974, and split some indenting into a separate commit. May move more indenting, need to see.

Anyways, hopefully should make the final commit easier to review :crossed_fingers:

sharnoff commented on neonvm/apis/neonvm/v1/virtualmachine_types.go:
```diff
@@ -558,6 +604,7 @@ func (vm *VirtualMachine) Cleanup() {
 	vm.Status.Node = ""
 	vm.Status.CPUs = nil
 	vm.Status.MemorySize = nil
+	vm.Status.MemoryProvider = nil
```
There's a couple reasons I went for this approach. This was my thought process, lmk what you think:

1. We need _some_ way to know how to scale the memory on the runner pod. This can either be a field in the spec or status.
2. We don't want to _need_ the control plane to set it; that way, we get control over rollout. So, as you mentioned -- either spec or status, but we have to auto-set _something_.
3. IMO: We should have an escape hatch, once virtio-mem is deployed. I.e., for a particular VM, we should be able to say "hey, please use DIMM slots instead". Eventually this should be a project/endpoint/compute setting (until we decide it's stable enough), and then we can remove all the old code. This belongs in the spec.
4. IMO: Once the VM is up and running, we _shouldn't_ allow changes to MemoryProvider setting. Actually, it needs to be _before_ the VM is up and running, to prevent race conditions. This can either be done with a status field (because special permissions are required), or with webhook update validation (i.e. only allow updates if pending and the podName wasn't set). We've had [issues around rollout of new fields + webhook validation in the past](https://www.notion.so/neondatabase/d66861b2d8a8444887282d8558428791?pvs=4#3ddf05468a8a44378b4c9931ad3dca95), so I'm hesitant to go that route.
5. IMO: _it would be nice_ if forcing VMs to pick up a change in MemoryProvider could just be done by deleting the pods. Otherwise we have to do object "delete + recreate with modified spec" (which is kind of a pain). I'm thinking of e.g. "oh no, there's an issue with virtio-mem, we need to switch out all affected runner pods". This requires a separate field in the status, if we also want to respect the spec.
6. IMO: _it would be nice_ if we could avoid neonvm-controller updating the spec (because I view "spec" as kind of "here's what the user asked", so to me it feels a little odd to update the spec. but... we have defaults anyways, so maybe this is analogous to that?)

From going through it again, I think it would be _fine_ to just use the spec, but I do also like having the status be separate. With (5) I'm probably just being paranoid.

WDYT?

sharnoff commented on neonvm/apis/neonvm/v1/virtualmachine_webhook.go:
```diff
@@ -187,6 +195,19 @@ func (r *VirtualMachine) ValidateUpdate(old runtime.Object) error {
 
+	// Only allow changing .spec.guest.memoryProvider if it was previously nil, and the new version
+	// is equal to .status.memoryProvider.
+	if !reflect.DeepEqual(before.Spec.Guest.MemoryProvider, r.Spec.Guest.MemoryProvider) {
```
Yeah fair enough. Done in 072fe6a130bb1884f79319ec68d0452d92935c4b. Won't squash until the discussion below is resolved, because IIUC we'll have to make this mutable if we put MemoryProvider in the spec, to allow neonvm-controller to set it

sharnoff commented:
Did backwards-compatibility testing locally (started a VM, upgraded the version, checked status was as expected, assuming that our existing e2e tests can take it from there). Found & fixed a bug.
Also checked rollback safety; yeah, seems fine :)
One interesting thing, posted it here: https://neondb.slack.com/archives/C039YKBRZB4/p1718950172991649

Omrigan commented on neonvm/controllers/vm_controller.go:
```diff
@@ -715,41 +729,71 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
+			// Maybe we're already using the amount we want?
+			// Update the status to reflect the current size - and if it matches goalTotalSize, ram
+			// is done.
+			currentTotalSize, err := QmpGetMemorySize(QmpAddr(vm))
```
I think this is a redundant check, we could get the current memory size from where it is set. 

But no worries, I'll clean it up in #943

sharnoff replied:
I _think_ in testing I actually saw this value change in between QMP queries in the same reconcile cycle. Not totally sure though...

Omrigan commented on neonvm/controllers/vm_controller.go:
```diff
@@ -715,41 +729,71 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
+
+			// Maybe we're already using the amount we want?
+			// Update the status to reflect the current size - and if it matches goalTotalSize, ram
+			// is done.
```
```suggestion
			// scaling is done.
```

Omrigan commented on neonvm/controllers/vm_controller.go:
```diff
@@ -812,6 +861,23 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
+	if p := vm.Spec.Guest.MemoryProvider; p != nil {
+		return *p
+	}
+	if p := vm.Status.MemoryProvider; p != nil {
```
Won't this prevent us from changing from one provider to the other by deleting the pod, because the old value be preserved through Status?

sharnoff replied:
the tl;dr on that is that we reset `.status.memoryProvider` in the same place we reset `.status.podName` (there's some method on `VirtualMachineStatus` that does it for us).

I tested it locally and it works as expected :)

Omrigan commented on neonvm/controllers/vm_qmp_queries.go:
```diff
@@ -333,6 +333,65 @@ type QMPRunner interface {
+	if vm.Spec.Guest.MemorySlots.Min == vm.Spec.Guest.MemorySlots.Max {
+		// if target size is non-zero even though min == max, something went very wrong
+		if targetVirtioMemSize != 0 {
+			panic(fmt.Errorf(
```
```suggestion
			panic(fmt.Sprintf(
```

Usually panic's arg is the string.

Omrigan commented on tests/e2e/autoscaling.virtio-mem/01-assert.yaml:
```diff
@@ -0,0 +1,23 @@
+kind: VirtualMachine
+metadata:
+  name: example
+status:
```
can we assert on the provider here?

sharnoff replied:
Done

Omrigan reviewed (approved):
Looks good overall

sharnoff commented on neonvm/controllers/vm_controller.go:
```diff
@@ -359,16 +366,23 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
 		vm.Status.Phase = vmv1.VmPending
 	case vmv1.VmPending:
-		// Generate runner pod name
+		// Generate runner pod name and set desired memory provider. After this, we know that
+		// MemoryProvider will be set, because earlier (above the 'switch vm.Status.Phase'), we set
+		// the MemoryProvider iff PodName was set but it wasn't.
```
Nice - I like that a lot better :)

sharnoff commented on neonvm/controllers/vm_controller.go:
```diff
@@ -722,38 +722,38 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
-			// firstly re-fetch VM
-			if err := r.Get(ctx, types.NamespacedName{Name: vm.Name, Namespace: vm.Namespace}, vm); err != nil {
-				log.Error(err, "Unable to re-fetch VirtualMachine")
+			realSlots, err := QmpSetMemorySlots(ctx, vm, targetSlotCount, r.Recorder)
```
Extracted in d2b8d897c917835790acb35b157e2833a6109d0d, replaced the indent commit with 1f05547e8c0f88e361229556fcd5b695a69b40d2 (splitting the first commit).
